index.rst

```diff
@@ -1,63 +1,830 @@
+The lab controller does this by generating ``/etc/passwd`` and ``/etc/group`` files and mounting them into the lab container over top of the files provided by the container image.
+
+The base ``/etc/passwd`` and ``/etc/group`` files are whatever minimal files are required to make the container work and provide reasonable human-readable usernames and groups for files present in the container.
+
+``/etc/passwd`` as mounted in the container has one added entry for the user.
+Their name, UID, and primary GID are taken from the user identity information associated with their token.
+The home directory is always ``/home/<username>`` and the shell is always ``/bin/bash``.
+
+``/etc/group`` as mounted in the container has an entry for each group in the user's group membership that has an associated GID.
+Groups without GIDs cannot be meaningfully represented in the ``/etc/group`` structure and are ignored.
+The user is added as a supplementary member of the group unless the GID of the group matches the user's primary GID.
+
+No ``/etc/shadow`` or ``/etc/gshadow`` files are mounted in the pod.
+The pod is always executed as the intended user and PAM should not be used or needed, so nothing should need or be able to read those files.
+
+The ``/etc/passwd`` and ``/etc/group`` files will be stored under ``passwd`` and ``group`` keys in a ``ConfigMap`` named ``nb-<username>-nss`` (from Name Service Switch, the name of the Linux subsystem that provides this type of user and group information), and mounted via the ``Pod`` specification.
+
+Mounts
+------
+
+The necessary volume mounts for a lab pod will be specified in the Helm configuration of the lab controller.
+At the least, this will include a mount definition for ``/home``, where user home directories must be located inside the pod.
+
+Secrets
+-------
+
+Each lab pod will have an associated ``Secret`` object named ``nb-<username>`` containing any required secrets.
+
+It will have at least one key, ``token``, which holds the notebook token for the user that is mounted into the pod and used to make API calls from the Notebook Aspect.
+This token is obtained from the ``POST`` request that creates the lab, via the ``X-Auth-Request-Token`` header added by Gafaelfawr.
+That route in the lab controller API will request a delegated notebook token.
+
+Additional secrets may be added via the Helm configuration of the lab controller.
+Each configured secret should be a reference to another ``Secret`` in the lab controller namespace and a key in that ``Secret`` object that should be added to the ``Secret`` object for each created pod.
+The secrets will be copied into the created ``Secret`` object during pod creation.
+
+One of those secrets may be tagged as a pull secret, in which case the required configuration to use it as a pull secret will also be added to the ``Pod`` specification.
+
+``NetworkPolicy``
```

Comment:
I realize why this is literal text but it would look better as "Network Policy".

How do we define "external IP addresses"?  If you just mean "not-RFC-1918 space" then we also (I think) need an exception for our storage systems (at least at the IDF I think these are addressed through non-public IP ranges).